addons/web/controllers/json2.py

```diff
@@ -0,0 +1,141 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import inspect
+import logging
+from collections.abc import Callable, Mapping, Sequence
+from http import HTTPStatus
+from typing import Any
+
+from werkzeug.exceptions import (
+    HTTPException,
+    BadRequest, NotFound, MethodNotAllowed, UnprocessableEntity,
+)
+
+from odoo import http
+from odoo.exceptions import UserError
+from odoo.http import request
+from odoo.models import BaseModel
+from odoo.service.model import get_public_method
+from odoo.tools import frozendict
+
+_logger = logging.getLogger(__name__)
+
+
+class Json2RpcDispatcher(http.Dispatcher):
+    routing_type = '/json/2/rpc'
+
+    @classmethod
+    def is_compatible_with(cls, request):
+        return request.httprequest.mimetype == 'application/json'
+
+    def dispatch(self, endpoint, args):
+        self.request.params = dict(args)
+        if self.request.httprequest.content_length:
+            try:
+                self.request.params |= self.request.get_json_data()
+            except ValueError as exc:
+                raise BadRequest(f"could not parse the body as json: {exc.args[0]}") from exc
+            except TypeError as exc:
+                # We don't need anything else than json object at the
+                # moment but we are open to discuss supporting the other
+                # data types. Reach us if you have a use case!
+                raise BadRequest("could not parse the body as a json object") from exc
```

Comment:
Nitpick. We don't need anything else and documenting it would be painful.
```suggestion
                raise BadRequest("could not parse the body, expecting a json object") from exc
```